FlowCal/plot.py

```diff
@@ -1058,21 +1141,33 @@ def scatter2d(data_list,
         plt.ylabel(data_plot.channels[1])
 
     # Set scale of axes
-    plt.gca().set_xscale(xscale)
-    plt.gca().set_yscale(yscale)
+    plt.gca().set_xscale(xscale, data=data_list, channel=channels[0])
+    plt.gca().set_yscale(yscale, data=data_list, channel=channels[1])
 
-    # Set plot limits if specified, else extract range from data_plot
+    # Set plot limits if specified, else extract range from data_list
     # Use data_plot.hist_bins with one single bin
-    if xlim is not None:
-        plt.xlim(xlim)
-    elif hasattr(data_plot, 'hist_bins') and \
-            hasattr(data_plot.hist_bins, '__call__'):
-        plt.xlim(data_plot.hist_bins(channels=0, nbins=1, scale=xscale))
-    if ylim is not None:
-        plt.ylim(ylim)
-    elif hasattr(data_plot, 'hist_bins') and \
-            hasattr(data_plot.hist_bins, '__call__'):
-        plt.ylim(data_plot.hist_bins(channels=1, nbins=1, scale=yscale))
+    if xlim is None:
+        xlim = [np.inf, -np.inf]
+        for data in data_list:
+            if hasattr(data, 'hist_bins') and \
+                    hasattr(data.hist_bins, '__call__'):
+                xlim_data = data.hist_bins(channels=channels[0],
+                                           nbins=1,
+                                           scale=xscale)
```

Comment:
Yet another way would be to make the range go from `1` to `$PnR` instead of `0` to `$PnR - 1`. We talked about this in #196, where I said the following:

> Assume that the range is `[0, $PnR]` and the resolution is `$PnR`. The current implementation of `hist_bins()` would then generate, by default, `$PnR` bins with weird spacing different than one. We could live with these weird bins, or we could make the resolution `$PnR + 1` with floating points, or we could have the range always be `[0, $PnR - 1]`.

> I choose the latter. So far we have two independent examples of floating point files, one of which blatantly disregards the standard's description of `$PnR` (i.e. has obviously a higher resolution ADC). In this file, events seem to be in `[0, $PnR]`, although there are negative numbers. Our other, more standard example does have channels with events in `[0, $PnR - 1]`, while using an ADC that clearly has the resolution specified by `PnR`, as I showed you before.

> Ultimately, the precise range for floating points may not be that important, given that the standard specifies that the file may contain numbers outside it. I think what I chose is the best option, but I don't think it's better by much.

However, if we need to plot in log scale, having the lower limit be zero is inconvenient, because `log(0)` is undefined. So one solution would be to change this to `[1, $PnR]`, and then we would be able to always call `.range` and log the result.